[libpas] PGM Crash Testing: Align allocation to page size

stwrt opened the pull request:
#### 69db44623766e6e27c87f20d7cb5add5220d59e9
<pre>
[libpas] PGM Crash Testing: Align allocation to page size
<a href="https://bugs.webkit.org/show_bug.cgi?id=276181">https://bugs.webkit.org/show_bug.cgi?id=276181</a>
<a href="https://rdar.apple.com/problem/131064444">rdar://problem/131064444</a>

Reviewed by David Kilzer.

PGM will align pages left or right, so by allocating the entire length of the page
it will give the same result whether left or right aligned. This will make testing guard pages reliable again.

* Source/JavaScriptCore/jsc.cpp:
(crashPGMUAF):
(crashPGMUpperGuardPage):
(crashPGMLowerGuardPage):

Canonical link: <a href="https://commits.webkit.org/280836@main">https://commits.webkit.org/280836@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/81b50aab0c5927a3485380f5e5e2a7d06484c829

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57750 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37078 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10226 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61372 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8195 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44714 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8383 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46789 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5810 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59780 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34801 "Found 2 new test failures: compositing/geometry/video-opacity-overlay.html fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49903 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27614 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31543 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7196 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7199 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/50842 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53495 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7466 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63054 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/56992 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1664 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7539 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54008 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1670 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49914 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54125 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12782 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1405 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/78753 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32907 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13056 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33993 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35077 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33738 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -3937,15 +3938,16 @@ static NEVER_INLINE void crashPGMUpperGuardPage()
+    size_t allocSize = getpagesize() * 10000;
+    char* result = static_cast<char*>(fastMalloc(allocSize));
     result = result - getpagesize();
     *result = 'a';
```
I think we should change this to write one byte before the start of the allocation instead of a whole page.  (This matches `crashPGMUpperGuardPage()` which writes one byte past the end of the allocation.)

I think subtracting 1 will work here:

```suggestion
    char* result = static_cast<char*>(fastMalloc(allocSize));
    result = result - 1;
    *result = 'a';
```

stwrt replied:
That sounds good. Will try that tomorrow.

ddkilzer reviewed (approved):
r=me

stwrt committed:
[libpas] PGM Crash Testing: Align allocation to page size
https://bugs.webkit.org/show_bug.cgi?id=276181
rdar://problem/131064444

Reviewed by David Kilzer.

PGM will align pages left or right, so by allocating the entire length of the page
it will give the same result whether left or right aligned. This will make testing guard pages reliable again.

* Source/JavaScriptCore/jsc.cpp:
(crashPGMUAF):
(crashPGMUpperGuardPage):
(crashPGMLowerGuardPage):

Canonical link: https://commits.webkit.org/280836@main
